README.md
# JWT Decode Full

[![Pub Version](https://img.shields.io/pub/v/jwt_decode_full?color=blueviolet)](https://pub.dev/packages/jwt_decode_full)

A JWT decoder that allows you to extract all the necessary data like the header, payload, expiration date and issued date from a JWT.

## Installation

Add `jwt_decode_full` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

<h1>Usage</h1>

```dart
  import 'package:jwt_decode_full/jwt_decode_full.dart';

  final jwtData = jwtDecode(myToken);

  print('header: ${jwtData.header}');
  print('payload: ${jwtData.payload}');
  print('isExpired: ${jwtData.isExpired}');
  print('issued date: ${jwtData.issuedAt}');
  print('expiration date: ${jwtData.expiration}');

  /// Result:
  ///
  /// header: {alg: PS256, typ: JWT}
  /// payload: {sub: 1234567890, name: John Doe, admin: true, iat: 1516239022, exp: 1517239022}
  /// isExpired: true
  /// issued date: 2018-01-18 02:30:22.000
  /// expiration date: 2018-01-29 16:17:02.000
```

For more detailed examples you can check out the [example project](https://github.com/dbilgin/jwt_decode_full/tree/master/example).


CHANGELOG.md
## 0.9.2

Fixed expiration calculation.

## 0.9.1

Fixed the readme import and title.

## 0.9.0

Main JWT decoding functionality for extracting:

- headers
- payload
- issuedAt
- expiration
- isExpired


test/jwt_decode_test.dart
import 'package:jwt_decode_full/jwt_decode_full.dart';
import 'package:test/test.dart';

const rs256 =
    'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWUsImlhdCI6MTUxNjIzOTAyMn0.NHVaYe26MbtOYhSKkoKYdFVomg4i8ZJd8_-RU8VNbftc4TSMb4bXP3l3YlNWACwyXPGffz5aXHc6lty1Y2t4SWRqGteragsVdZufDn5BlnJl9pdR_kdVFUsra2rWKEofkZeIC4yWytE58sMIihvo9H1ScmmVwBcQP6XETqYd0aSHp1gOa9RdUPDvoXQ5oqygTqVtxaDr6wUFKrKItgBMzWIdNZ6y7O9E0DhEPTbE9rfBo6KTFsHAZnMg4k68CDp2woYIaXbmYTWcvbzIuHO7_37GT79XdIwkm95QJ7hYC9RiwrV7mesbY4PAahERJawntho0my942XheVLmGwLMBkQ';
const hs512 =
    'eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWUsImlhdCI6MTUxNjIzOTAyMn0.VFb0qJ1LRg_4ujbZoRMXnVkUgiuKq5KxWqNdbKq_G9Vvz-S1zZa9LPxtHWKa64zDl2ofkT8F6jBt_K4riU-fPg';
const ps256WithExpiration =
    'eyJhbGciOiJQUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWUsImlhdCI6MTUxNjIzOTAyMiwiZXhwIjoxNTE3MjM5MDIyfQ.bYr7VB4XfnNzxzdMeRJpP5IK9xTMmHPzeLhFh4ndnpgXE4_pE1iGx_q3i7eMqbeWVMK-BzxkCN8KNKxRib1XaVRzD0MTRaebFXDL5zbPjTU0ZzP24PThzsu4RuqTrTBifbW04wMar_4jdjImNYUnK61dYOXZoOtEOCeZdcNMrf_jaMcSJ8B0bQvbMf07X6aKZ9RN9YqSn51De5GgqlwUq_22hac5gDpfg3kpeah1eq1uzJXze78dXPOZUtSga3bFLnCrdQ-QwRivJ6F_8VHlCIHvFW8HqRixSeZWtWwyPYFo93dpm0jkpISXwSCfoxor1-5Bnivnqtccaost6LoohQ';

void main() {
  group('decode', () {
    test("rs256 token", () {
      final data = jwtDecode(rs256);
      expect(data.header['alg'], 'RS256');
      expect(data.payload['name'], 'John Doe');
      expect(data.isExpired, isNull);
      expect(data.expiration, isNull);
      expect(data.issuedAt, DateTime.fromMillisecondsSinceEpoch(1516239022000));
    });

    test("hs512 token", () {
      final data = jwtDecode(hs512);
      expect(data.header['alg'], 'HS512');
      expect(data.payload['name'], 'John Doe');
      expect(data.isExpired, isNull);
      expect(data.expiration, isNull);
      expect(data.issuedAt, DateTime.fromMillisecondsSinceEpoch(1516239022000));
    });

    test("ps256 token with expiration", () {
      final data = jwtDecode(ps256WithExpiration);
      expect(data.header['alg'], 'PS256');
      expect(data.payload['name'], 'John Doe');
      expect(data.isExpired, true);
      expect(
        data.expiration,
        DateTime.fromMillisecondsSinceEpoch(1517239022000),
      );
      expect(data.issuedAt, DateTime.fromMillisecondsSinceEpoch(1516239022000));
    });
  });

  group('invalid token', () {
    test("throws format exception without dots", () {
      expect(() => jwtDecode(rs256.replaceAll('.', '')), throwsFormatException);
    });

    test("throws format exception", () {
      expect(() => jwtDecode('in.val.id'), throwsFormatException);
    });
  });
}


lib/src/jwt_data.dart
/// `JWTData` is a class that contains the header, payload of a JWT.
/// By using the `isExpired`, `issuedAt`, and `expiration` getters, you can
/// easily extract the calculated results for these properties.
class JWTData {
  JWTData({required this.header, required this.payload});
  final Map<String, dynamic> header;
  final Map<String, dynamic> payload;

  bool? get isExpired {
    final exp = payload['exp'] as int?;
    return exp == null
        ? null
        : exp < DateTime.now().millisecondsSinceEpoch / 1000;
  }

  DateTime? get issuedAt {
    final iat = payload['iat'] as int?;
    return iat == null ? null : DateTime.fromMillisecondsSinceEpoch(iat * 1000);
  }

  DateTime? get expiration {
    final exp = payload['exp'] as int?;
    return exp == null ? null : DateTime.fromMillisecondsSinceEpoch(exp * 1000);
  }
}


lib/src/jwt_decode.dart
import 'dart:convert';

import 'jwt_data.dart';

/// `jwtDecode` provides a way to decode different types of JWTs and
/// extract their payloads, headers, expiration dates, and issued dates and
/// whether they are expired or not.
///
/// You can use this function to decode a JWT and return a [JWTData] object:
/// ```dart
/// final jwtData = jwtDecode(myToken);
///
/// print('header: ${jwtData.header}');
/// print('payload: ${jwtData.payload}');
/// print('isExpired: ${jwtData.isExpired}');
/// print('issued date: ${jwtData.issuedAt}');
/// print('expiration date: ${jwtData.expiration}');
/// ```
JWTData jwtDecode(String token) {
  final splitToken = token.split('.');
  if (splitToken.length != 3) {
    throw const FormatException('Invalid token');
  }

  final headerEncoded = splitToken[0];
  final payloadEncoded = splitToken[1];

  final normalizedHeader = base64.normalize(headerEncoded);
  final headerString = utf8.decode(base64.decode(normalizedHeader));
  final decodedHeader = jsonDecode(headerString) as Map<String, dynamic>;

  final normalizedPayload = base64.normalize(payloadEncoded);
  final payloadString = utf8.decode(base64.decode(normalizedPayload));
  final decodedPayload = jsonDecode(payloadString) as Map<String, dynamic>;

  return JWTData(header: decodedHeader, payload: decodedPayload);
}


example/jwt_decode_full_example.dart
import 'package:jwt_decode_full/jwt_decode_full.dart';

const myToken =
    'eyJhbGciOiJQUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWUsImlhdCI6MTUxNjIzOTAyMiwiZXhwIjoxNTE3MjM5MDIyfQ.bYr7VB4XfnNzxzdMeRJpP5IK9xTMmHPzeLhFh4ndnpgXE4_pE1iGx_q3i7eMqbeWVMK-BzxkCN8KNKxRib1XaVRzD0MTRaebFXDL5zbPjTU0ZzP24PThzsu4RuqTrTBifbW04wMar_4jdjImNYUnK61dYOXZoOtEOCeZdcNMrf_jaMcSJ8B0bQvbMf07X6aKZ9RN9YqSn51De5GgqlwUq_22hac5gDpfg3kpeah1eq1uzJXze78dXPOZUtSga3bFLnCrdQ-QwRivJ6F_8VHlCIHvFW8HqRixSeZWtWwyPYFo93dpm0jkpISXwSCfoxor1-5Bnivnqtccaost6LoohQ';

void main() {
  final jwtData = jwtDecode(myToken);

  print('header: ${jwtData.header}');
  print('payload: ${jwtData.payload}');
  print('isExpired: ${jwtData.isExpired}');
  print('issued date: ${jwtData.issuedAt}');
  print('expiration date: ${jwtData.expiration}');

  /// Result:
  ///
  /// header: {alg: PS256, typ: JWT}
  /// payload: {sub: 1234567890, name: John Doe, admin: true, iat: 1516239022, exp: 1517239022}
  /// isExpired: true
  /// issued date: 2018-01-18 02:30:22.000
  /// expiration date: 2018-01-29 16:17:02.000
}
